<!doctype html>
<html>
    <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">

        <title>Async Shizzle</title>

        <link rel="stylesheet" href="css/reveal.css">
        <link rel="stylesheet" href="css/theme/blood.css">

        <!-- Theme used for syntax highlighting of code -->
        <link rel="stylesheet" href="lib/css/zenburn.css">

        <!-- Printing and PDF exports -->
        <script>
            var link = document.createElement( 'link' );
            link.rel = 'stylesheet';
            link.type = 'text/css';
            link.href = window.location.search.match( /print-pdf/gi ) ? 'css/print/pdf.css' : 'css/print/paper.css';
            document.getElementsByTagName( 'head' )[0].appendChild( link );
        </script>
    </head>
    <body>
        <div class="reveal">
            <div class="slides">
                <section>
                    <h1>Async JS</h1>
                    <ul>
                        <li>Callback</li>
                        <li>Promise</li>
                        <li>async await</li>
                    </ul>
                </section>
                <section>
                    <h1>Callback</h1>
                </section>
                <section>
                    <h3>function</h3>
                    <pre><code data-trim data-noescape>
                      function answerQuestion (question, cb) {
                        setTimeout(function() { // sum poor demonstration delay

                          if ( Math.random() > .5 ) {
                            return cb( new Error('Something went wrong') )
                          }

                          return cb( null, 42 )

                         }, 500)
                      }
                    </code></pre>
                </section>
                <section>
                    <h3>Single</h3>
                    <pre><code data-trim data-noescape>
                        function callback (err, value) {
                          if ( err ) {
                            return handleError(err)
                          }

                          console.log(value)
                        }

                        answerQuestion('Whats the purpose of life?', callback)
                    </code></pre>
                </section>
                <section>
                    <h2>Vorteile:</h2>
                    <ul>
                        <li>Funktioniert in jedem Browser</li>
                        <li>Benötigt keine zusätzliche Objekte wie z.B. Promise</li>
                    </ul>
                </section>
                <section>
                    <h2>Nachteil: </h2>
                    <h3>Callbackhell</h3>
                </section>
                <section>
                    <h4>multiple series</h4>
                    <pre><code data-trim data-noescape>
                        const answer = answerQuestion

                        answer('Whats the purpose of life?', function ( err, answer1 ) {
                          if ( err ) return handleError(err)

                          answer('What is 21 times 2?', function ( err, answer2 ) {
                            if ( err ) return handleError(err)

                             answer('What is 84 divided by 2?', function ( err, answer3 ) {
                               if ( err ) return handleError(err)

                               doSthWithAllAnswers([ answer1, answer2, answer3 ])
                             })
                          })
                        })
                    </code></pre>
                </section>
                <section>
                    <h4>multiple series</h4>
                    <pre><code data-trim data-noescape>
                        const answer = answerQuestion
                        const answers = []

                        function answerStuff() {
                          answer('Whats the purpose of life?', function ( err, answer1 ) {
                            if ( err ) return handleError(err)
                             answers.push( answer1 )
                             answer2()
                          })
                        }

                        function answer2() {
                          answer('What is 21 times 2?', function ( err, answer2 ) {
                            if ( err ) return handleError(err)

                            answers.push( answer2 )
                            answer3()
                          })
                        }

                        function answer3() {
                          answer('What is 84 divided by 2?', function ( err, answer3 ) {
                            if ( err ) return handleError(err)

                            answers.push( answer3 )
                            doSthWithAllAnswers( answers )
                          })
                        }

                        answerStuff()
                    </code></pre>
                </section>
                <section>
                    <h3>multiple series (async.js)</h3>
                    <pre><code data-trim data-noescape>
                        async.series([
                          callback => answerQuestion('Whats the purpose of life?', callback),
                          callback => answerQuestion('What is 21 times 2?', callback),
                          callback => answerQuestion('What is 84 divided by 2?', callback),
                        ], function(err, answers) {
                          if ( err ) return handleError(err)

                          doSthWithAllAnswers(answers)
                        })
                    </code></pre>
                    <span>http://caolan.github.io/async/</span>
                </section>
                <section>
                    <h3>multiple parallel</h3>
                    <pre><code data-trim data-noescape>
                        let i = 0
                        let answers = []
                        let occurredError = null
                        function callback( err, result ) {
                          if ( occurredError ) {
                            return
                          } else if ( err ) {
                            occuredError = err
                            return handleError( occuredError )
                          }

                          i++;
                          answers.push( result )
                          if ( i === 3 ) {
                            doSthWithAllAnswers(answers)
                          }
                        }
                        answerQuestion('Whats the purpose of life?', callback)
                        answerQuestion('What is 21 times 2?', callback)
                        answerQuestion('What is 84 divided by 2?', callback)
                    </code></pre>
                </section>
                <section>
                    <h3>multiple parallel (async.js)</h3>
                    <pre><code data-trim data-noescape>
                        const answer = answerQuestion
                        async.parallel([
                          callback => answer('Whats the purpose of life?', callback),
                          callback => answer('What is 21 times 2?', callback),
                          callback => answer('What is 84 divided by 2?', callback),
                        ], ( err, answers ) => {
                          if ( err ) {
                            return handleError( err )
                          }
                          doSthWithAllAnswers( answers )
                        })
                    </code></pre>
                </section>
                <section>
                    <h1>Promise</h1>
                </section>
                <section>
                    <h3>function</h3>
                    <pre><code data-trim data-noescape>
                        function answer(question) {
                          return new Promise((resolve, reject) => {
                            setTimeout(() => { // sum poor demonstration delay

                            if ( Math.random() > .5 ) {
                              return reject( new Error('Something went wrong') )
                            }

                            resolve( 42 )

                            }, 500)
                          })
                        }
                    </code></pre>
                </section>
                <section>
                    <h3>single</h3>
                    <pre><code data-trim data-noescape>
                        answerQuestion('Whats the purpose of life?').then( answer => {
                          console.log( answer )
                        }).catch( err => {
                          handleError(err)
                        })
                    </code></pre>
                </section>
                <section>
                    <h3>multiple series</h3>
                    <pre><code data-trim data-noescape>
                        const answers = []
                        answerQuestion('Whats the purpose of life?').then( result => {
                          answers.push( result )
                          return answerQuestion('What is 21 times 2?')
                        }).then( result => {
                          answers.push( result )
                          return answerQuestion('What is 84 divided by 2?')
                        }).then( result => {
                          answers.push( result )
                          doSthWithAllAnswers( answers )
                        }).catch( err => {
                          handleError( err )
                        })
                    </code></pre>
                </section>
                <section>
                    <h3>multiple parallel</h3>
                    <pre><code data-trim data-noescape>
                        Promise.all([
                          answerQuestion('Whats the purpose of life?'),
                          answerQuestion('What is 21 times 2?'),
                          answerQuestion('What is 84 divided by 2?'),
                        ]).then( answers => {
                          doSthWithAllAnswers( answers )
                        }).catch( err => {
                          handleError( err )
                        })
                    </code></pre>
                </section>
                <section>
                    <h2>Vorteil:</h2>
                    <ul>
                        <li>Besseres Error Handling</li>
                    </ul>
                </section>
                <section>
                    <h2>Nachteil:</h2>
                    <ul>
                        <li>ES2015 -> Language Feature</li>
                        <li>Polyfill benötigt</li>
                    </ul>
                </section>
                <section>
                    <h2>Warum Promise verwenden?</h2>
                    <ul>
                        <li>Language Feature (ES2015)</li>
                        <li>Macht Libraries wie async.js (in den meisten Fällen) überflüssig</li>
                    </ul>
                </section>
                <section>
                    <h3>Browser APIs</h3>
                    <ul>
                        <li><a href="https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API">fetch</a></li>
                        <li><a href="https://developer.mozilla.org/en-US/docs/Web/API/Web_Animations_API">Web Animations</a></li>
                        <li><a href="https://developer.mozilla.org/de/docs/Web/API/Service_Worker_API">Service Worker</a></li>
                    </ul>
                </section>
                <section>
                    <h1>Async Await</h1>
                </section>
                <section>
                    <h3>function</h3>
                    <pre><code data-trim data-noescape>
                        function answerQuestion(question) {
                          return new Promise((resolve, reject) => {
                            setTimeout(() => { // sum poor demonstration delay

                            if ( Math.random() > .5 ) {
                              return reject( new Error('Something went wrong') )
                            }

                            resolve( 42 )

                            }, 500)
                          })
                        }
                    </code></pre>
                </section>
                <section>
                    <h3>single</h3>
                    <pre><code data-trim data-noescape>
                        const answer = answerQuestion
                        async function doStuff() {
                          try {
                            const result = await answer('Whats the purpose of life?')
                            console.log( result )
                          } catch( err ) {
                            handleError( err )
                          }
                        }

                        doStuff()
                    </code></pre>
                </section>
                <section>
                    <h3>multiple series</h3>
                    <pre><code data-trim data-noescape>
                        async function answerMyQuestions() {
                          try {
                            const answer1 = await answerQuestion('Whats the purpose of life?')
                            const answer2 = await answerQuestion('What is 21 times 2?')
                            const answer3 = await answerQuestion('What is 84 divided by 2?')

                            doSthWithAllAnswers( [ answer1, answer2, answer3 ])
                          } catch ( err ) {
                            handleError( err )
                          }
                        }

                        answerMyQuestions()
                    </code></pre>
                </section>
                <section>
                    <h3>multiple parallel</h3>
                    <pre><code data-trim data-noescape>
                        async function answerMyQuestions() {
                          try {
                            const answers = await Promise.all([
                              answerQuestion('Whats the purpose of life?'),
                              answerQuestion('What is 21 times 2?'),
                              answerQuestion('What is 84 divided by 2?'),
                            ])

                            doSthWithAllAnswers( answers )
                          } catch ( err ) {
                            handleError( err )
                          }
                        }

                        answerMyQuestions()
                    </code></pre>
                </section>
                <section>
                    <h2>Vorteile</h2>
                    <ul>
                        <li>Async-Code sieht aus wie Sync-Code</li>
                        <li>Error Handling mit try catch</li>
                        <li>Weniger Code</li>
                    </ul>
                </section>
                <section>
                    <h2>Nachteile</h2>
                    <ul>
                        <li>Kein Polyfill möglich</li>
                        <li>Code muss transformiert werden</li>
                    </ul>
                </section>
                <section>
                    <h2>Transform Beispiel (ES2015 env)</h2>
                    <pre><code data-trim data-noescape>
                        let answerMyQuestions = (() => {
                          var _ref = _asyncToGenerator(function* () {
                            try {
                              const answer1 = yield answerMyQuestionAsync('Whats the purpose of life?');
                              const answer2 = yield answerQuestionAsync('What is 21 times 2?');
                              const answer3 = yield answerQuestionAsync('What is 84 divided by 2?');

                              doSthWithAllAnswers([answer1, answer2, answer3]);
                            } catch (err) {
                              handleError(err);
                            }
                          });

                          return function answerMyQuestions() {
                            return _ref.apply(this, arguments);
                          };
                        })();

                        function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

                        answerMyQuestions();
                    </code></pre>
                </section>
                <section>
                    <h2>Transform Beispiel (ES5 env)</h2>
                    <pre><code data-trim data-noescape>
                        var answerMyQuestions = function () {
                          var _ref = _asyncToGenerator(regeneratorRuntime.mark(function _callee() {
                            var answer1, answer2, answer3;
                            return regeneratorRuntime.wrap(function _callee$(_context) {
                              while (1) {
                                switch (_context.prev = _context.next) {
                                  case 0:
                                    _context.prev = 0;
                                    _context.next = 3;
                                    return answerMyQuestionAsync('Whats the purpose of life?');

                                  case 3:
                                    answer1 = _context.sent;
                                    _context.next = 6;
                                    return answerQuestionAsync('What is 21 times 2?');

                                  case 6:
                                    answer2 = _context.sent;
                                    _context.next = 9;
                                    return answerQuestionAsync('What is 84 divided by 2?');

                                  case 9:
                                    answer3 = _context.sent;


                                    doSthWithAllAnswers([answer1, answer2, answer3]);
                                    _context.next = 16;
                                    break;

                                  case 13:
                                    _context.prev = 13;
                                    _context.t0 = _context['catch'](0);

                                    handleError(_context.t0);

                                  case 16:
                                  case 'end':
                                    return _context.stop();
                                }
                              }
                            }, _callee, this, [[0, 13]]);
                          }));

                          return function answerMyQuestions() {
                            return _ref.apply(this, arguments);
                          };
                        }();

                        function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

                        answerMyQuestions();
                    </code></pre>
                </section>
                <section>
                    <h2><a href="https://github.com/tc39/proposal-dynamic-import">Dynamic Import</a></h2>
                    <ul>
                        <li>Vorschlag für das asynchrone Laden von Modulen</li>
                        <li>Stage 3 (TC39)</li>
                    </ul>
                </section>
                <section>
                    <h3>Beispiel</h3>
                    <pre><code data-trim data-noescape>
                        import('./my-module').then( myModule => {
                          myModule.doSth()
                        })

                        async function main() {
                          const myModule = await import('./my-module')
                          myModule.doSth()
                        }
                        main()
                    </code></pre>
                </section>
            </div>
        </div>

        <script src="lib/js/head.min.js"></script>
        <script src="js/reveal.js"></script>

        <script>
            // More info https://github.com/hakimel/reveal.js#configuration
            Reveal.initialize({
                history: true,

                // More info https://github.com/hakimel/reveal.js#dependencies
                dependencies: [
                    { src: 'plugin/markdown/marked.js' },
                    { src: 'plugin/markdown/markdown.js' },
                    { src: 'plugin/notes/notes.js', async: true },
                    { src: 'plugin/highlight/highlight.js', async: true, callback: function() { hljs.initHighlightingOnLoad(); } }
                ]
            });
        </script>
    </body>
</html>
